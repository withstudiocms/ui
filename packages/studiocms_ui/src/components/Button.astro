---
import type { HTMLTag, Polymorphic } from 'astro/types';
import type { StudioCMSColorway } from '../utils/colors';

type Props<As extends HTMLTag = 'button'> = Omit<Polymorphic<{ as: As }>, 'as'> & {
	as?: As;
	size?: 'sm' | 'md' | 'lg';
	fullWidth?: boolean;
	color?: StudioCMSColorway;
	variant?: 'solid' | 'outlined' | 'flat';
	class?: string;
	disabled?: boolean;
};

const {
	size = 'md',
	fullWidth = false,
	color = 'default',
	variant = 'default',
	as: As = 'button',
	class: className,
	disabled = false,
	...props
} = Astro.props;
---

<As
  class="sui-button"
  class:list={[
    fullWidth && "full",
    disabled && "disabled",
    className,
    size,
    color,
    variant,
  ]}
  disabled={disabled}
	tabindex={0}
  {...props}
>
  <slot name="start-content" />
  <slot />
  <slot name="end-content" />
</As>
<style>
	/* Default Styles */
	.sui-button {
		display: flex;
		flex-direction: row;
		align-items: center;
		text-align: center;

		position: relative;

		/* Move gaps & radii to spacings.css later on */
		gap: 0.5rem;

		outline: none;
		border: none;

		font-weight: 400;

		border-radius: 0.5rem;

		transition: all 0.15s ease;

		cursor: pointer;

		/* Default colorway */

		background-color: hsl(var(--default-base));
		border-color: hsl(var(--border));
		color: hsl(var(--text-muted));

		min-width: fit-content;

		will-change: transform;

		text-decoration: none;
	}

	.sui-button.disabled {
		pointer-events: none;
		opacity: 0.5;
	}

	.sui-button.outlined {
		border-width: 1px;
		border-style: solid;
	}

	.sui-button:hover {
		background-color: hsl(var(--default-hover));
	}

	.sui-button:active {
		background-color: hsl(var(--default-active));
		transform: scale(0.95);
	}

	.sui-button.flat {
		background: hsla(var(--default-flat));
	}

	.sui-button.flat:hover {
		background: hsla(var(--default-flat-hover));
	}

	.sui-button.flat:active {
		background: hsla(var(--default-flat-active));
	}

	/* Sizes */
	.sui-button.sm {
		height: 32px;
		padding: 0.5rem 0.75rem;

		font-size: 0.825em;
	}

	.sui-button.md {
		height: 40px;
		padding: 0.5rem 0.75rem;

		font-size: 0.875em;
	}

	.sui-button.lg {
		height: 48px;
		padding: 0.75rem 1rem;

		font-size: 1em;
	}

	/* Primary Colorway */
	.sui-button.primary {
		background-color: hsl(var(--primary-base));
		border-color: hsl(var(--primary-base));
		color: hsl(var(--text-inverted));
	}

	.sui-button.primary:hover {
		background-color: hsl(var(--primary-hover));
	}

	.sui-button.primary:active {
		background-color: hsl(var(--primary-active));
	}

	.sui-button.primary.outlined {
		background-color: hsl(var(--default-base));
		color: hsl(var(--text-normal));
	}

	.sui-button.primary.outlined:hover {
		background-color: hsl(var(--default-hover));
	}

	.sui-button.primary.outlined:active {
		background-color: hsl(var(--default-active));
	}

	.sui-button.primary.flat {
		background-color: hsla(var(--primary-flat));
		color: hsl(259, 96%, 79%);
	}

	.sui-button.primary.flat:hover {
		background-color: hsla(var(--primary-flat-hover));
	}

	.sui-button.primary.flat:active {
		background-color: hsla(var(--primary-flat-active));
	}

	/* Success Colorway */
	.sui-button.success {
		background-color: hsl(var(--success-base));
		border-color: hsl(var(--success-base));
		color: hsl(var(--text-dark));
	}

	.sui-button.success:hover {
		background-color: hsl(var(--success-hover));
	}

	.sui-button.success:active {
		background-color: hsl(var(--success-active));
	}

	.sui-button.success.outlined {
		background-color: hsl(var(--default-base));
		color: hsl(var(--text-normal));
	}

	.sui-button.success.outlined:hover {
		background-color: hsl(var(--default-hover));
	}

	.sui-button.success.outlined:active {
		background-color: hsl(var(--default-active));
	}

	.sui-button.success.flat {
		background-color: hsla(var(--success-flat));
		color: hsl(var(--success-base));
	}

	.sui-button.success.flat:hover {
		background-color: hsla(var(--success-flat-hover));
	}

	.sui-button.success.flat:active {
		background-color: hsla(var(--success-flat-active));
	}

	/* Warning Colorway */
	.sui-button.warning {
		background-color: hsl(var(--warning-base));
		border-color: hsl(var(--warning-base));
		color: hsl(var(--text-dark));
	}

	.sui-button.warning:hover {
		background-color: hsl(var(--warning-hover));
	}

	.sui-button.warning:active {
		background-color: hsl(var(--warning-active));
	}

	.sui-button.warning.outlined {
		background-color: hsl(var(--default-base));
		color: hsl(var(--text-normal));
	}

	.sui-button.warning.outlined:hover {
		background-color: hsl(var(--default-hover));
	}

	.sui-button.warning.outlined:active {
		background-color: hsl(var(--default-active));
	}

	.sui-button.warning.flat {
		background-color: hsla(var(--warning-flat));
		color: hsl(var(--warning-base));
	}

	.sui-button.warning.flat:hover {
		background-color: hsla(var(--warning-flat-hover));
	}

	.sui-button.warning.flat:active {
		background-color: hsla(var(--warning-flat-active));
	}

	/* Danger Colorway */
	.sui-button.danger {
		background-color: hsl(var(--danger-base));
		border-color: hsl(var(--danger-base));
		color: hsl(var(--text-light));
	}

	.sui-button.danger:hover {
		background-color: hsl(var(--danger-hover));
	}

	.sui-button.danger:active {
		background-color: hsl(var(--danger-active));
	}

	.sui-button.danger.outlined {
		background-color: hsl(var(--default-base));
		color: hsl(var(--text-normal));
	}

	.sui-button.danger.outlined:hover {
		background-color: hsl(var(--default-hover));
	}

	.sui-button.danger.outlined:active {
		background-color: hsl(var(--default-active));
	}

	.sui-button.danger.flat {
		background-color: hsla(var(--danger-flat));
		color: hsl(337, 92%, 71%);
	}

	.sui-button.danger.flat:hover {
		background-color: hsla(var(--danger-flat-hover));
	}

	.sui-button.danger.flat:active {
		background-color: hsla(var(--danger-flat-active));
	}

	.sui-button.full {
		width: 100%;
		justify-content: center;
	}

	[data-theme="light"] {
		.sui-button.primary.flat {
			color: hsl(259, 84%, 45%);
		}

		.sui-button.success.flat {
			background-color: hsl(var(--success-base), 0.2);
			color: hsl(143, 59%, 20%);
		}

		.sui-button.success.flat:hover {
			background-color: hsla(var(--success-flat-hover));
		}

		.sui-button.success.flat:active {
			background-color: hsla(var(--success-flat-active));
		}

		.sui-button.warning.flat {
			background-color: hsla(var(--warning-base), 0.2);
			color: hsl(48, 78%, 20%);
		}

		.sui-button.warning.flat:hover {
			background-color: hsla(var(--warning-base), 0.25);
		}

		.sui-button.warning.flat:active {
			background-color: hsla(var(--warning-base), 0.4);
		}

		.sui-button.danger.flat {
			color: hsl(339, 97%, 29%);
		}
	}
</style>
