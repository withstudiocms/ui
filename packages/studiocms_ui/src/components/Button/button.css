/* Default Styles */
.sui-button {
  display: flex;
  flex-direction: row;
  align-items: center;
  text-align: center;

  position: relative;

  /* Move gaps & radii to spacings.css later on */
  gap: 0.5rem;

  outline: none;
  border: none;

  font-weight: 400;

  border-radius: var(--radius-md);

  transition: transform 0.15s, background-color 0.15s, border-color 0.15s, color 0.15s;
  transition-timing-function: ease;

  cursor: pointer;

  /* Default colorway */

  background-color: hsl(var(--default-base));
  border-color: hsl(var(--border));
  color: hsl(var(--text-normal));

  min-width: fit-content;

  will-change: transform;

  text-decoration: none;
}

.sui-button.disabled {
  pointer-events: none;
  opacity: 0.5;
}

.sui-button.outlined {
  border-width: 1px;
  border-style: solid;
}

.sui-button.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.outlined:active {
  background-color: hsl(var(--default-hover));
}

.sui-button:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button:active {
  background-color: hsl(var(--default-active));
  transform: scale(0.95);
}

.sui-button.flat {
  background: hsla(var(--default-flat));
}

.sui-button.flat:hover {
  background: hsla(var(--default-flat-hover));
}

.sui-button.flat:active {
  background: hsla(var(--default-flat-active));
}

/* Sizes */
.sui-button.sm {
  height: 32px;
  padding: 0.5rem 0.75rem;

  font-size: 0.825em;
}

.sui-button.md {
  height: 40px;
  padding: 0.5rem 0.75rem;

  font-size: 0.875em;
}

.sui-button.lg {
  height: 48px;
  padding: 0.75rem 1rem;

  font-size: 1em;
}

.sui-button:focus-visible {
  outline: 2px solid hsl(var(--text-normal));
  outline-offset: 2px;
}

/* Primary Colorway */
.sui-button.primary {
  background-color: hsl(var(--primary-base));
  border-color: hsl(var(--primary-base));
  color: hsl(var(--text-inverted));
}

.sui-button.primary:hover {
  background-color: hsl(var(--primary-hover));
}

.sui-button.primary:active {
  background-color: hsl(var(--primary-active));
}

.sui-button.primary.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.primary.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.primary.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.primary.flat {
  background-color: hsla(var(--primary-flat));
  color: hsl(259, 96%, 79%);
}

.sui-button.primary.flat:hover {
  background-color: hsla(var(--primary-flat-hover));
}

.sui-button.primary.flat:active {
  background-color: hsla(var(--primary-flat-active));
}

/* Success Colorway */
.sui-button.success {
  background-color: hsl(var(--success-base));
  border-color: hsl(var(--success-base));
  color: hsl(var(--text-dark));
}

.sui-button.success:hover {
  background-color: hsl(var(--success-hover));
}

.sui-button.success:active {
  background-color: hsl(var(--success-active));
}

.sui-button.success.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.success.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.success.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.success.flat {
  background-color: hsla(var(--success-flat));
  color: hsl(var(--success-base));
}

.sui-button.success.flat:hover {
  background-color: hsla(var(--success-flat-hover));
}

.sui-button.success.flat:active {
  background-color: hsla(var(--success-flat-active));
}

/* Warning Colorway */
.sui-button.warning {
  background-color: hsl(var(--warning-base));
  border-color: hsl(var(--warning-base));
  color: hsl(var(--text-dark));
}

.sui-button.warning:hover {
  background-color: hsl(var(--warning-hover));
}

.sui-button.warning:active {
  background-color: hsl(var(--warning-active));
}

.sui-button.warning.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.warning.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.warning.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.warning.flat {
  background-color: hsla(var(--warning-flat));
  color: hsl(var(--warning-base));
}

.sui-button.warning.flat:hover {
  background-color: hsla(var(--warning-flat-hover));
}

.sui-button.warning.flat:active {
  background-color: hsla(var(--warning-flat-active));
}

/* Danger Colorway */
.sui-button.danger {
  background-color: hsl(var(--danger-base));
  border-color: hsl(var(--danger-base));
  color: hsl(var(--text-light));
}

.sui-button.danger:hover {
  background-color: hsl(var(--danger-hover));
}

.sui-button.danger:active {
  background-color: hsl(var(--danger-active));
}

.sui-button.danger.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.danger.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.danger.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.danger.flat {
  background-color: hsla(var(--danger-flat));
  color: hsl(337, 92%, 71%);
}

.sui-button.danger.flat:hover {
  background-color: hsla(var(--danger-flat-hover));
}

.sui-button.danger.flat:active {
  background-color: hsla(var(--danger-flat-active));
}

/* Info Colorway */
.sui-button.info {
  background-color: hsl(var(--info-base));
  border-color: hsl(var(--info-base));
  color: hsl(var(--text-light));
}

.sui-button.info:hover {
  background-color: hsl(var(--info-hover));
}

.sui-button.info:active {
  background-color: hsl(var(--info-active));
}

.sui-button.info.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.info.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.info.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.info.flat {
  background-color: hsla(var(--info-flat));
  color: hsl(var(--info-base));
}

.sui-button.info.flat:hover {
  background-color: hsla(var(--info-flat-hover));
}

.sui-button.info.flat:active {
  background-color: hsla(var(--info-flat-active));
}

/* Mono Colorway */
.sui-button.mono {
  background-color: hsl(var(--mono-base));
  border-color: hsl(var(--mono-base));
  color: hsl(var(--text-inverted));
}

.sui-button.mono:hover {
  background-color: hsl(var(--mono-hover));
}

.sui-button.mono:active {
  background-color: hsl(var(--mono-active));
}

.sui-button.mono.outlined {
  background-color: hsl(var(--default-base));
  color: hsl(var(--text-normal));
}

.sui-button.mono.outlined:hover {
  background-color: hsl(var(--default-hover));
}

.sui-button.mono.outlined:active {
  background-color: hsl(var(--default-active));
}

.sui-button.mono.flat {
  background-color: hsla(var(--mono-flat));
  color: hsl(var(--mono-base));
}

.sui-button.mono.flat:hover {
  background-color: hsla(var(--mono-flat-hover));
}

.sui-button.mono.flat:active {
  background-color: hsla(var(--mono-flat-active));
}

.sui-button.full {
  width: 100%;
  justify-content: center;
}

[data-theme="light"] {
  .sui-button.primary.flat {
    color: hsl(259, 84%, 45%);
  }

  .sui-button.success.flat {
    background-color: hsl(var(--success-base), 0.2);
    color: hsl(143, 59%, 20%);
  }

  .sui-button.success.flat:hover {
    background-color: hsla(var(--success-flat-hover));
  }

  .sui-button.success.flat:active {
    background-color: hsla(var(--success-flat-active));
  }

  .sui-button.warning.flat {
    background-color: hsla(var(--warning-base), 0.2);
    color: hsl(48, 78%, 20%);
  }

  .sui-button.warning.flat:hover {
    background-color: hsla(var(--warning-base), 0.25);
  }

  .sui-button.warning.flat:active {
    background-color: hsla(var(--warning-base), 0.4);
  }

  .sui-button.danger.flat {
    color: hsl(339, 97%, 29%);
  }
}
