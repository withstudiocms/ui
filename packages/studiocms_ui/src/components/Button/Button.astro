---
import type { HTMLTag, Polymorphic } from 'astro/types';
import type { StudioCMSColorway } from '../../utils/colors.js';
import './button.css';

/**
 * Props for the button component.
 */
type Props<As extends HTMLTag = 'button'> = Omit<Polymorphic<{ as: As }>, 'as'> & {
	/**
	 * The polymorphic component to render the button as. Defaults to `button`.
	 */
	as?: As;
	/**
	 * The size of the button. Defaults to `md`.
	 */
	size?: 'sm' | 'md' | 'lg';
	/**
	 * Whether the button should be full width. Defaults to `false`.
	 */
	fullWidth?: boolean;
	/**
	 * The colorway of the button. Defaults to `default`.
	 */
	color?: StudioCMSColorway;
	/**
	 * The variant of the button. Defaults to `solid`.
	 */
	variant?: 'solid' | 'outlined' | 'flat';
	/**
	 * Whether the button is disabled. Defaults to `false`.
	 */
	disabled?: boolean;
	/**
	 * An optional href to use for the button. If provided, the button will render as an anchor tag.
	 */
	href?: string;
};

export type { Props };

const {
	size = 'md',
	fullWidth = false,
	color = 'default',
	variant = 'solid',
	disabled = false,
	...props
} = Astro.props;

let As: HTMLTag = 'button';

if ('href' in props) {
	As = 'a';
} else {
	As = props.as || 'button';
}
---

<As
  class="sui-button"
  class:list={[
    fullWidth && "full",
    disabled && "disabled",
    size,
    color,
    variant,
  ]}
  disabled={disabled}
  tabindex={0}
  {...props}>
  <slot name="start-content" />
  <slot />
  <slot name="end-content" />
</As>
