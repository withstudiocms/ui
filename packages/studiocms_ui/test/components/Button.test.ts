import type { ComponentProps } from 'astro/types';
import { describe, expect } from 'vitest';
import Button from '../../src/components/Button/Button.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Button Component', () => {
	test('renders Button component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Button, 'Button', { slots: { default: 'Test Button' } });
		expect(result).toMatchSnapshot();
	});

	test('renders Button component with different sizes', async ({ renderComponent }) => {
		const sizeVariants: ComponentProps<typeof Button>[] = [
			{ size: 'sm' },
			{ size: 'md' },
			{ size: 'lg' },
		];

		for (const props of sizeVariants) {
			const result = await renderComponent(Button, 'Button', {
				props,
				slots: { default: 'Test Button' },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Button component with different colors', async ({ renderComponent }) => {
		const colorVariants: ComponentProps<typeof Button>[] = [
			{ color: 'primary' },
			{ color: 'default' },
			{ color: 'success' },
			{ color: 'danger' },
			{ color: 'warning' },
			{ color: 'info' },
			{ color: 'mono' },
		];

		for (const props of colorVariants) {
			const result = await renderComponent(Button, 'Button', {
				props,
				slots: { default: 'Test Button' },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Button component with different variants', async ({ renderComponent }) => {
		const variantVariants: ComponentProps<typeof Button>[] = [
			{ variant: 'solid' },
			{ variant: 'outlined' },
			{ variant: 'flat' },
		];

		for (const props of variantVariants) {
			const result = await renderComponent(Button, 'Button', {
				props,
				slots: { default: 'Test Button' },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Button component with miscellaneous props', async ({ renderComponent }) => {
		const miscVariants: ComponentProps<typeof Button>[] = [
			{ fullWidth: true },
			{ disabled: true },
			{ as: 'a', href: 'https://example.com' },
		];

		for (const props of miscVariants) {
			const result = await renderComponent(Button, 'Button', {
				props,
				slots: { default: 'Test Button' },
			});
			expect(result).toMatchSnapshot();
		}
	});

	test('renders Button with different slots', async ({ renderComponent }) => {
		const slots: { name: string; content: string }[] = [
			{ name: 'default', content: 'Default Slot' },
			{ name: 'start-content', content: 'Prefix Slot' },
			{ name: 'end-content', content: 'Suffix Slot' },
		];

		for (const slot of slots) {
			const result = await renderComponent(Button, 'Button', {
				slots: { [slot.name]: slot.content },
			});
			expect(result).toMatchSnapshot();
		}
	});
});
