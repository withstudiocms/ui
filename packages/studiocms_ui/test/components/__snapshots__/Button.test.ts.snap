// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > renders Button component correctly 1`] = `"<button tabindex="0" class="sui-button md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 1`] = `"<button tabindex="0" class="sui-button md primary solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 2`] = `"<button tabindex="0" class="sui-button md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 3`] = `"<button tabindex="0" class="sui-button md success solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 4`] = `"<button tabindex="0" class="sui-button md danger solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 5`] = `"<button tabindex="0" class="sui-button md warning solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 6`] = `"<button tabindex="0" class="sui-button md info solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different colors 7`] = `"<button tabindex="0" class="sui-button md mono solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different sizes 1`] = `"<button tabindex="0" class="sui-button sm default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different sizes 2`] = `"<button tabindex="0" class="sui-button md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different sizes 3`] = `"<button tabindex="0" class="sui-button lg default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different variants 1`] = `"<button tabindex="0" class="sui-button md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different variants 2`] = `"<button tabindex="0" class="sui-button md default outlined">  Test Button  </button>"`;

exports[`Button Component > renders Button component with different variants 3`] = `"<button tabindex="0" class="sui-button md default flat">  Test Button  </button>"`;

exports[`Button Component > renders Button component with miscellaneous props 1`] = `"<button tabindex="0" class="sui-button full md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with miscellaneous props 2`] = `"<button disabled tabindex="0" class="sui-button disabled md default solid">  Test Button  </button>"`;

exports[`Button Component > renders Button component with miscellaneous props 3`] = `"<a tabindex="0" as="a" href="https://example.com" class="sui-button md default solid">  Test Button  </a>"`;

exports[`Button Component > renders Button with different slots 1`] = `"<button tabindex="0" class="sui-button md default solid">  Default Slot  </button>"`;

exports[`Button Component > renders Button with different slots 2`] = `"<button tabindex="0" class="sui-button md default solid"> Prefix Slot   </button>"`;

exports[`Button Component > renders Button with different slots 3`] = `"<button tabindex="0" class="sui-button md default solid">   Suffix Slot </button>"`;
